tools/protoprint/protoprint_test.py

```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env python3
+
+import argparse
+import logging
+import os
+import pathlib
+import re
+import subprocess
+import sys
+import tarfile
+import tempfile
+
+from tools.protoprint.test_data import data as test_data
+from tools.run_command import run_command
+
+from envoy_repo import PATH
+
+
+def path_and_filename(label):
+    """Retrieve actual path and filename from bazel label
+
+    Args:
+        label: bazel label to specify target proto.
+
+    Returns:
+        actual path and filename
+    """
+    if label.startswith('/'):
+        label = label.replace('//', '/', 1)
+    elif label.startswith('@'):
+        label = re.sub(r'@.*/', '/', label)
+    else:
+        return label
+    label = label.replace(":", "/")
+    splitted_label = label.split('/')
+    return '/'.join(splitted_label[:len(splitted_label) - 1])[1:], splitted_label[-1]
+
+
+def golden_proto_file(tmp, path, filename, version):
+    """Retrieve golden proto file path. In general, those are placed in tools/testdata/protoxform.
+
+    Args:
+        path: target proto path
+        filename: target proto filename
+        version: api version to specify target golden proto filename
+
+    Returns:
+        actual golden proto absolute path
+    """
+
+    gold_dir = tmp.joinpath("golden")
+    filename = f"{filename}.{version}.gold"
+    return gold_dir.joinpath(filename).absolute()
+
+
+def proto_print(protoprint, descriptor, src, dst):
+    """Pretty-print FileDescriptorProto to a destination file.
+
+    Args:
+        src: source path for FileDescriptorProto.
+        dst: destination path for formatted proto.
+    """
+    print('proto_print %s -> %s' % (src, dst))
+    subprocess.check_call(
+        [protoprint, src, dst, descriptor, './tools/testdata/protoxform/TEST_API_VERSION'])
+
+
+def result_proto_file(tmp, protoprint, descriptor, path, filename, version):
+    """Retrieve result proto file path. In general, those are placed in bazel artifacts.
+
+    Args:
+        cmd: fix or freeze?
+        path: target proto path
+        tmp: temporary directory.
+        filename: target proto filename
+        version: api version to specify target result proto filename
+
+    Returns:
+        actual result proto absolute path
+    """
+
+    pkg_dir = tmp.joinpath("xformed")
+    base = pkg_dir.joinpath("fix_protos").joinpath(path).joinpath(f"{filename}.{version}.proto")
+    dst = pathlib.Path(filename).absolute()
+    proto_print(protoprint, descriptor, str(base.absolute()), str(dst))
+    return dst
+
+
+def diff(result_file, golden_file):
+    """Execute diff command with unified form
+
+    Args:
+        result_file: result proto file
+        golden_file: golden proto file
+
+    Returns:
+        output and status code
+    """
+    command = 'diff -u '
+    command += str(result_file) + ' '
+    command += str(golden_file)
```

Comment:
Prefer a join on a list here..

Reply:
will swap to an f-string - should say that i havent made too much of an attempt to cleanup this on the basis of minimum change